File: Database/sqliteConnector.py
import sqlite3
from Models.parliamentGroup import ParliamentGroup

connection = sqlite3.connect("parliament.db")

# Allows for query traceability
#connection.set_trace_callback(print)

def initDb():
    cursor = connection.cursor()

    cursor.execute("SELECT name FROM sqlite_master WHERE type='table' AND name='Initiatives';")

    if cursor.fetchall() == []:
        print("Creating tables")
        cursor.execute("CREATE TABLE Initiatives(id INTEGER, title TEXT, text TEXT, PRIMARY KEY(id))")
        cursor.execute("CREATE TABLE Votes(id INTEGER, initiativeId INTEGER, fase TEXT, result TEXT, details TEXT, date TEXT, PRIMARY KEY(id), FOREIGN KEY(initiativeId) REFERENCES Initiatives(id) )")
        cursor.execute("CREATE TABLE ParliamentGroups(acronym TEXT, name TEXT, PRIMARY KEY(acronym))")

    connection.commit()

def getParliamentGroups():
    cursor = connection.cursor()
    cursor.execute("SELECT * FROM ParliamentGroups")
    parliamentGroups = cursor.fetchall()
    
    cursor.close()

    return list(map(lambda g: ParliamentGroup(g[0], g[1]), parliamentGroups))

def getInitiatives():
    cursor = connection.cursor()
    cursor.execute("SELECT * FROM Initiatives")
    initiatives = cursor.fetchall()

    cursor.close()

    return initiatives

def insertParliamentGroups(parliamentGroups):
    cursor = connection.cursor()
    parliamentGroupsTuple = map(lambda g: (g.acronym, g.name), parliamentGroups)
    cursor.executemany("INSERT INTO ParliamentGroups VALUES(?, ?)", parliamentGroupsTuple)
    cursor.close()

    connection.commit()

def insertInitiatives(initiatives):

    cursor = connection.cursor()

    cursor.execute("DELETE FROM Initiatives")
    cursor.execute("DELETE FROM Votes")

    connection.commit()

    initiativesToInsert = []
    votesToInsert = []

    for initiative in initiatives:
        initiativeTuple = (initiative.id, initiative.title, initiative.text)
        votesTuple = map(lambda v: (int(v.id), int(initiative.id),
                                    v.phase if v.phase is not None else "",
                                    v.result if v.result is not None else "",
                                    v.votingDetails if v.votingDetails is not None else "",
                                    v.date if v.date is not None else "")
                                    , initiative.votes)

        initiativesToInsert.append(initiativeTuple)
        votesToInsert.extend(list(votesTuple))

    cursor.executemany("INSERT INTO Initiatives VALUES(?, ?, ?)", initiativesToInsert)
    cursor.executemany("INSERT INTO Votes VALUES(?, ?, ?, ?, ?, ?)", votesToInsert)

    cursor.close()

    connection.commit()

def close():
    connection.close()

File: API/Database/sqliteConnector.ts
const { AsyncDatabase } = require("promised-sqlite3");
import dbConnector from '../Interfaces/dbConnector.js';

export default class SqliteConnector implements dbConnector{

    dbPath:string;
    db:any;

    constructor(dbPath: string){
        this.dbPath = dbPath;
    }

    async getParliamentGroups(): Promise<any[]> {
        await this.connect();
        const query = "SELECT * FROM ParliamentGroups";

        const result = await this.db.all(query);

        let parliamentGroups: any = [];

        result.forEach((row: { acronym: any; name: any; }) => {
            parliamentGroups.push({
                acronym: row.acronym,
                name: row.name
            })
        });

        await this.close();

        return parliamentGroups;
    }

    async getParliamentGroupByAcronym(acronym: string): Promise<any> {
        await this.connect();

        const query = `SELECT * FROM ParliamentGroups WHERE acronym ='${acronym}'`;
        const result = await this.db.get(query);

        await this.close();

        return result;
    }

    async connect(){
        if(!this.db)
            this.db = await AsyncDatabase.open(this.dbPath);  
    }

    async close(){
        if(this.db)
            await this.db.close();
    }
}